File: backend/src/main/java/ru/bowling/bowlingapp/DTO/ServiceHistoryDTO.java
package ru.bowling.bowlingapp.DTO;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import jakarta.validation.constraints.*;
import java.time.LocalDateTime;
import java.util.List;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class ServiceHistoryDTO {

    private Long serviceId;

    @NotNull(message = "Club ID is required")
    private Long clubId;
    private String clubName;

    private Long equipmentId;
    private String equipmentName;

    @Min(value = 1, message = "Lane number must be greater than 0")
    private Integer laneNumber;

    @NotNull(message = "Service type is required")
    private String serviceType;

    private LocalDateTime serviceDate;

    @NotBlank(message = "Description is required")
    @Size(max = 5000, message = "Description cannot exceed 5000 characters")
    private String description;

    @Size(max = 2000, message = "Parts replaced description cannot exceed 2000 characters")
    private String partsReplaced;

    @DecimalMin(value = "0.0", message = "Labor hours must be non-negative")
    private Double laborHours;

    private Double totalCost;

    @NotNull(message = "Performed by mechanic ID is required")
    private Long performedByMechanicId;
    private String performedByMechanicName;

    private Long supervisedByUserId;
    private String supervisedByUserName;

    private LocalDateTime nextServiceDue;
    private LocalDateTime warrantyUntil;

    @Size(max = 2000, message = "Service notes cannot exceed 2000 characters")
    private String serviceNotes;

    private String performanceMetrics; // JSON string
    private List<String> photos;
    private List<String> documents;

    private LocalDateTime createdDate;
    private Long createdBy;
    private String createdByName;

    private List<ServiceHistoryPartDTO> partsUsed;
}

File: frontend/lib/models/service_history_dto.dart
import 'service_history_part_dto.dart';

class ServiceHistoryDto {
  final int? serviceId;
  final int clubId;
  final String? clubName;
  final int? equipmentId;
  final String? equipmentName;
  final int? laneNumber;
  final String serviceType;
  final DateTime? serviceDate;
  final String description;
  final String? partsReplaced;
  final double? laborHours;
  final double? totalCost;
  final int performedByMechanicId;
  final String? performedByMechanicName;
  final int? supervisedByUserId;
  final String? supervisedByUserName;
  final DateTime? nextServiceDue;
  final DateTime? warrantyUntil;
  final String? serviceNotes;
  final String? performanceMetrics;
  final List<String>? photos;
  final List<String>? documents;
  final DateTime? createdDate;
  final int? createdBy;
  final String? createdByName;
  final List<ServiceHistoryPartDto>? partsUsed;

  ServiceHistoryDto({
    this.serviceId,
    required this.clubId,
    this.clubName,
    this.equipmentId,
    this.equipmentName,
    this.laneNumber,
    required this.serviceType,
    this.serviceDate,
    required this.description,
    this.partsReplaced,
    this.laborHours,
    this.totalCost,
    required this.performedByMechanicId,
    this.performedByMechanicName,
    this.supervisedByUserId,
    this.supervisedByUserName,
    this.nextServiceDue,
    this.warrantyUntil,
    this.serviceNotes,
    this.performanceMetrics,
    this.photos,
    this.documents,
    this.createdDate,
    this.createdBy,
    this.createdByName,
    this.partsUsed,
  });

  factory ServiceHistoryDto.fromJson(Map<String, dynamic> json) {
    DateTime? d(dynamic v) => (v is String && v.isNotEmpty) ? DateTime.parse(v) : null;
    double? f(dynamic v) => v == null ? null : (v as num).toDouble();
    List<String>? ls(dynamic v) => v is List ? v.map((e) => e.toString()).toList() : null;
    return ServiceHistoryDto(
      serviceId: (json['serviceId'] as num?)?.toInt(),
      clubId: (json['clubId'] as num).toInt(),
      clubName: json['clubName'] as String?,
      equipmentId: (json['equipmentId'] as num?)?.toInt(),
      equipmentName: json['equipmentName'] as String?,
      laneNumber: (json['laneNumber'] as num?)?.toInt(),
      serviceType: json['serviceType'] as String,
      serviceDate: d(json['serviceDate']),
      description: json['description'] as String,
      partsReplaced: json['partsReplaced'] as String?,
      laborHours: f(json['laborHours']),
      totalCost: f(json['totalCost']),
      performedByMechanicId: (json['performedByMechanicId'] as num).toInt(),
      performedByMechanicName: json['performedByMechanicName'] as String?,
      supervisedByUserId: (json['supervisedByUserId'] as num?)?.toInt(),
      supervisedByUserName: json['supervisedByUserName'] as String?,
      nextServiceDue: d(json['nextServiceDue']),
      warrantyUntil: d(json['warrantyUntil']),
      serviceNotes: json['serviceNotes'] as String?,
      performanceMetrics: json['performanceMetrics'] as String?,
      photos: ls(json['photos']),
      documents: ls(json['documents']),
      createdDate: d(json['createdDate']),
      createdBy: (json['createdBy'] as num?)?.toInt(),
      createdByName: json['createdByName'] as String?,
      partsUsed: (json['partsUsed'] as List?)
          ?.map((e) => ServiceHistoryPartDto.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
        'serviceId': serviceId,
        'clubId': clubId,
        'clubName': clubName,
        'equipmentId': equipmentId,
        'equipmentName': equipmentName,
        'laneNumber': laneNumber,
        'serviceType': serviceType,
        'serviceDate': serviceDate?.toIso8601String(),
        'description': description,
        'partsReplaced': partsReplaced,
        'laborHours': laborHours,
        'totalCost': totalCost,
        'performedByMechanicId': performedByMechanicId,
        'performedByMechanicName': performedByMechanicName,
        'supervisedByUserId': supervisedByUserId,
        'supervisedByUserName': supervisedByUserName,
        'nextServiceDue': nextServiceDue?.toIso8601String(),
        'warrantyUntil': warrantyUntil?.toIso8601String(),
        'serviceNotes': serviceNotes,
        'performanceMetrics': performanceMetrics,
        'photos': photos,
        'documents': documents,
        'createdDate': createdDate?.toIso8601String(),
        'createdBy': createdBy,
        'createdByName': createdByName,
        'partsUsed': partsUsed?.map((e) => e.toJson()).toList(),
      };
}
